File: backend/src/main/java/tn/esprit/backend/Services/InvestmentHoldingService.java
package tn.esprit.backend.Services;

import tn.esprit.backend.DTO.CheckoutSessionResponse;
import tn.esprit.backend.DTO.ConfirmCheckoutRequest;
import tn.esprit.backend.DTO.CreateCheckoutSessionRequest;
import tn.esprit.backend.DTO.CreateMilestoneRequest;
import tn.esprit.backend.DTO.InvestmentHoldingMilestoneResponse;
import tn.esprit.backend.DTO.InvestmentHoldingResponse;

public interface InvestmentHoldingService {
    CheckoutSessionResponse createCheckoutSession(String requestId, CreateCheckoutSessionRequest request, RequestActor actor);
    InvestmentHoldingResponse confirmCheckout(ConfirmCheckoutRequest request);
    InvestmentHoldingResponse getHolding(String holdingId, RequestActor actor);
    InvestmentHoldingResponse getHoldingByRequestId(String requestId, RequestActor actor);
    InvestmentHoldingResponse requestRelease(String holdingId, RequestActor actor);
    InvestmentHoldingResponse releaseFunds(String holdingId, RequestActor actor);
    InvestmentHoldingResponse dispute(String holdingId, RequestActor actor);
    InvestmentHoldingResponse refund(String holdingId, RequestActor actor);
    InvestmentHoldingMilestoneResponse createMilestone(String holdingId, CreateMilestoneRequest request, RequestActor actor);
    InvestmentHoldingMilestoneResponse validateMilestone(String milestoneId, RequestActor actor);
    InvestmentHoldingMilestoneResponse releaseMilestone(String milestoneId, RequestActor actor);
}

File: frontend/src/app/features/investment/services/investment-holding.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { apiOrigin } from '../../../core/api-origin';
import {
  CheckoutSessionResponse,
  ConfirmCheckoutPayload,
  CreateHoldingPayload,
  CreateMilestonePayload,
  HoldingRole,
  InvestmentHolding,
  InvestmentHoldingMilestone,
} from '../models/investment-holding.model';

@Injectable({
  providedIn: 'root'
})
export class InvestmentHoldingService {
  private readonly baseUrl = `${apiOrigin()}/api/investments`;
  private readonly adminUrl = `${apiOrigin()}/api/admin/investments/holding`;

  constructor(private http: HttpClient) {}

  createCheckoutSession(requestId: string, payload: CreateHoldingPayload, userId: string, role: HoldingRole): Observable<CheckoutSessionResponse> {
    return this.http.post<CheckoutSessionResponse>(
      `${this.baseUrl}/${requestId}/checkout-session`,
      payload,
      { headers: this.buildHeaders(userId, role) }
    );
  }

  getHolding(holdingId: string, userId: string, role: HoldingRole): Observable<InvestmentHolding> {
    return this.http.get<InvestmentHolding>(
      `${this.baseUrl}/holding/${holdingId}`,
      { headers: this.buildHeaders(userId, role) }
    );
  }

  getHoldingByRequestId(requestId: string, userId: string, role: HoldingRole): Observable<InvestmentHolding> {
    return this.http.get<InvestmentHolding>(
      `${this.baseUrl}/request/${requestId}/holding`,
      { headers: this.buildHeaders(userId, role) }
    );
  }

  requestRelease(holdingId: string, userId: string, role: HoldingRole): Observable<InvestmentHolding> {
    return this.http.post<InvestmentHolding>(
      `${this.baseUrl}/holding/${holdingId}/request-release`,
      {},
      { headers: this.buildHeaders(userId, role) }
    );
  }

  dispute(holdingId: string, userId: string, role: HoldingRole): Observable<InvestmentHolding> {
    return this.http.post<InvestmentHolding>(
      `${this.baseUrl}/holding/${holdingId}/dispute`,
      {},
      { headers: this.buildHeaders(userId, role) }
    );
  }

  releaseFunds(holdingId: string, userId: string): Observable<InvestmentHolding> {
    return this.http.post<InvestmentHolding>(
      `${this.adminUrl}/${holdingId}/release`,
      {},
      { headers: this.buildHeaders(userId, 'ADMIN') }
    );
  }

  refund(holdingId: string, userId: string): Observable<InvestmentHolding> {
    return this.http.post<InvestmentHolding>(
      `${this.adminUrl}/${holdingId}/refund`,
      {},
      { headers: this.buildHeaders(userId, 'ADMIN') }
    );
  }

  createMilestone(holdingId: string, payload: CreateMilestonePayload, userId: string, role: HoldingRole): Observable<InvestmentHoldingMilestone> {
    return this.http.post<InvestmentHoldingMilestone>(
      `${this.baseUrl}/holding/${holdingId}/milestones`,
      payload,
      { headers: this.buildHeaders(userId, role) }
    );
  }

  validateMilestone(milestoneId: string, userId: string, role: HoldingRole): Observable<InvestmentHoldingMilestone> {
    return this.http.patch<InvestmentHoldingMilestone>(
      `${this.baseUrl}/holding/milestones/${milestoneId}/validate`,
      {},
      { headers: this.buildHeaders(userId, role) }
    );
  }

  releaseMilestone(milestoneId: string, userId: string): Observable<InvestmentHoldingMilestone> {
    return this.http.post<InvestmentHoldingMilestone>(
      `${this.adminUrl}/milestones/${milestoneId}/release`,
      {},
      { headers: this.buildHeaders(userId, 'ADMIN') }
    );
  }

  confirmCheckout(payload: ConfirmCheckoutPayload): Observable<InvestmentHolding> {
    return this.http.post<InvestmentHolding>(
      `${this.baseUrl}/confirm-checkout`,
      payload
    );
  }

  private buildHeaders(userId: string, role: HoldingRole): HttpHeaders {
    return new HttpHeaders({
      'X-User-Id': userId,
      'X-User-Role': role,
    });
  }
}
